src/html-transform.ts

```diff
@@ -0,0 +1,256 @@
+      options.isEntryPoint && !options.requireJsUrl) {
+    throw new Error(
+        'requireJsUrl is required when ' +
+        'js.transformEsModulesToAmd and isEntryPoint are true.');
+  }
+  if (options.js && options.js.moduleResolution === 'node' &&
+      !options.js.filePath) {
+    throw new Error('Cannot perform node module resolution without filePath.');
+  }
+
+  const document = parse5.parse(html);
+  const allScripts = dom5.queryAll(document, isJsScriptNode);
+
+  let wctScript;
+  const moduleScripts = [];
+  const otherScripts = [];
+  let shouldTransformEsModuleToAmd =
+      options.js && options.js.transformEsModulesToAmd;
+
+  for (const script of allScripts) {
+    if (dom5.getAttribute(script, 'type') === 'module') {
+      moduleScripts.push(script);
+    } else {
+      otherScripts.push(script);
+    }
+
+    if (dom5.hasAttribute(script, 'nomodule')) {
+      // Assume that if this document has a nomodule script, the author is
+      // already handling browsers that don't support modules, and we don't need
+      // to transform them (even if the configuration was set).
+      shouldTransformEsModuleToAmd = false;
+    }
+
+    const src = dom5.getAttribute(script, 'src') || '';
+    if (src.includes('web-component-tester/browser.js') ||
+        src.includes('wct-browser-legacy/browser.js')) {
+      wctScript = script;
+    }
+  }
+
+  // Script tags that aren't getting the AMD module treatment.
+  const standardTransformScripts = [...otherScripts];
+  let requireJsScript;
+
+  if (shouldTransformEsModuleToAmd) {
+    transformEsModulesToAmd(moduleScripts, options.js);
+
+    if (options.isEntryPoint && moduleScripts.length > 0) {
+      // We need RequireJS to load the AMD modules we are declaring. Inject the
+      // dependency as late as possible (just before the first module is
+      // declared) because there may be some UMD dependencies that we want to
+      // continue to load in global mode instead of AMD mode (which is detected
+      // by the presence of the `require` global).
+      const fragment = parse5.parseFragment(
+          `<script src="${options.requireJsUrl}"></script>\n`);
+      requireJsScript = fragment.childNodes![0];
+      const firstModule = moduleScripts[0];
+      dom5.insertBefore(firstModule.parentNode!, firstModule, fragment);
+    }
+
+  } else {
+    standardTransformScripts.push(...moduleScripts);
+  }
+
+  for (const script of standardTransformScripts) {
+    if (dom5.hasAttribute(script, 'src')) {
+      continue;
+    }
+    dom5.setTextContent(
+        script,
+        jsTransform(
+            dom5.getTextContent(script),
+            {...options.js, transformEsModulesToAmd: false}));
+  }
+
+  if (wctScript && requireJsScript) {
+    addWctTimingHack(wctScript, requireJsScript);
+  }
+
+  return parse5.serialize(document);
+}
+
+function transformEsModulesToAmd(
+    moduleScripts: dom5.Node[], jsOptions: JsTransformOptions|undefined) {
```

Comment:
can we just pass all module + defer scripts here and get the correct total ordering?

Reply:
Added defer scripts to the module chain and added tests.